### PR TITLE
Add 2.12.0 Blog post.

### PR DESCRIPTION
**Pull request by apilloud**
This adds the blog post for the 2.12.0 release.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: website/src/_posts/2019-04-19-beam-2.12.0.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+### I/Os
+
+* Implement a BigQuery custom sink for Python SDK.
+* Add support to specify a query in CassandraIO for Java SDK.
+
+### New Features / Improvements
+
+* Enable Bundle Finalization in Python SDK.
+* Add support to the Java SDK harness to merge windows.
```
Comment (tweise) [inline website/src/_posts/2019-04-19-beam-2.12.0.md]: I think we should mention `GenerateSequence` as first cross-language transform: https://issues.apache.org/jira/browse/BEAM-6730 - it may also be good to clarify what we have as of 2.12 in the ticket @mxm 

@aaltay is there anything that should be mentioned here regarding Python3 ?

Comment (mxm) [reply website/src/_posts/2019-04-19-beam-2.12.0.md]: Agree, we could mention cross-language transforms as a new feature.  

For [Flink](https://jira.apache.org/jira/issues/?jql=project%20%3D%20BEAM%20AND%20fixVersion%20%3D%202.12.0%20AND%20component%20%3D%20runner-flink), I'd also mention Kafka Sink EOS support.

File: website/src/_posts/2019-04-19-beam-2.12.0.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+We are happy to present the new 2.12.0 release of Beam. This release includes both improvements and new functionality.
+See the [download page]({{ site.baseurl }}/get-started/downloads/#2120-2019-04-19) for this release.<!--more-->
+For more information on changes in 2.12.0, check out the
+[detailed release notes](https://jira.apache.org/jira/secure/ReleaseNote.jspa?projectId=12319527&version=12344944).
+
+## Highlights
+
+### I/Os
+
+* Implement a BigQuery custom sink for Python SDK.
```
Comment (aaltay) [inline website/src/_posts/2019-04-19-beam-2.12.0.md]: Could we reword this to make it similar with other feature. "Added support for a BigQuery custom sink for Python SDK".

@pabloem could you add a JIRA link or another sentence here on what it means for the users and how they could use it?

File: website/src/_posts/2019-04-19-beam-2.12.0.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+* Add support to specify a query in CassandraIO for Java SDK.
+
+### New Features / Improvements
+
+* Enable Bundle Finalization in Python SDK.
```
Comment (aaltay) [inline website/src/_posts/2019-04-19-beam-2.12.0.md]: `Enable Bundle Finalization in Python SDK.` -> `Enable Bundle Finalization in Python SDK for portable runners.`

File: website/src/_posts/2019-04-19-beam-2.12.0.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+* Add support to the Java SDK harness to merge windows.
+
```
Comment (aaltay) [inline website/src/_posts/2019-04-19-beam-2.12.0.md]: Suggestion: "Added a dead letter queue to Python streaming BigQuery sink."

File: website/src/_posts/2019-04-19-beam-2.12.0.md (hunk 1 above)
Comment (aaltay) [reply website/src/_posts/2019-04-19-beam-2.12.0.md]: Python 3, things I am aware of
- bug fixes related to serialization. No need to mention this separately.
- py 3.6 and 3.7 workloads enabled -- still more experimental than 3.5, worth mentioning.
- BQ and Avro, was there a default change related to fastavro since last release?
- Does datastore io change happened?

@tvalentyn for the questions.

Comment (tweise) [reply website/src/_posts/2019-04-19-beam-2.12.0.md]: > Agree, we could mention cross-language transforms as a new feature.
> 

@mxm I would suggest we clarify what exactly the user can expect to work in 2.12 here and also on the JIRA ticket: https://issues.apache.org/jira/browse/BEAM-6730

Comment (tvalentyn) [reply website/src/_posts/2019-04-19-beam-2.12.0.md]: BQ, Avro, Datastore changes didn't make it into 2.12.
The 3.6, 3.7 change is specific to Dataflow runner only. We could say:
"Beam 2.12 supports starting Dataflow pipelines under Python 3.6, 3.7, however 3.5 remains the only recommended minor version for Dataflow runner. In addition to [announced 2.11 limitations](https://issues.apache.org/jira/browse/BEAM-1251?focusedCommentId=16789854&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16789854), Beam typehint annotations are currently not supported on Python >= 3.6."

Comment (mxm) [reply website/src/_posts/2019-04-19-beam-2.12.0.md]: @tweise I've added a summary in the JIRA issue. It is too long for the release notes but perhaps we could include something like this:

```
- Adds experimental support for cross-language transforms, 
  please see [BEAM-6730](https://issues.apache.org/jira/browse/BEAM-6730) 
```

```
- Adds support in the Flink Runner for exactly-once Writes with KafkaIO
```

**Comment by apilloud** [comment]
I originally sourced these from the release notes page, so it would be good to keep those JIRA issues named well in the future.

**Comment by apilloud** [comment]
Updated with feedback.

**Comment by apilloud** [comment]
https://apache-beam-website-pull-requests.storage.googleapis.com/8314/blog/2019/04/25/beam-2.12.0.html
R: @melap
